src/go/util/url_util.go

```diff
@@ -38,7 +38,22 @@ const (
 )
 
 var (
-	pathParamMatcherRegexp = regexp.MustCompile(`{[^{}]+}`)
+	// Various hacky regular expressions to match a subset of the http template syntax.
+	// Replace segments with single wildcards: /v1/books/*
+	singleWildcardMatcher = regexp.MustCompile(`/\*`)
+	// Replace segments with double wildcards: /v1/**
+	doubleWildcardMatcher = regexp.MustCompile(`/\*\*`)
+	// Replace any path templates: /v1/books/{book_id}
+	pathParamMatcher = regexp.MustCompile(`/{[^{}]+}`)
+	// Replace path templates with double wildcards: /v1/{name=**}
+	pathParamDoubleWildcardMatcher = regexp.MustCompile(`/{[^{}]+=\*\*}`)
+
+	// Common regex forms that emulate http template syntax.
+	// Matches 1 or more segments of any character except '/'.
+	singleWildcardReplacementRegex = `/[^\/]+`
+	// Matches any character or no characters at all.
+	// Note this also allows paths without a '/' suffix.
+	doubleWildcardReplacementRegex = `.*`
```

Comment:
I looked at the definition of url_template rules:  Its syntax:

// ### Path template syntax
//
//     Template = "/" Segments [ Verb ] ;
//     Segments = Segment { "/" Segment } ;
//     Segment  = "*" | "**" | LITERAL | Variable ;
//     Variable = "{" FieldPath [ "=" Segments ] "}" ;
//     FieldPath = IDENT { "." IDENT } ;
//     Verb     = ":" LITERAL ;

It has following forms
*
**
{x} which is same as {x=*}
{x=**}
{x=/foo/bar*}
{x=/foo/*/bar/**}

We could not support last two forms now.   

I am not sure how easy to support them.

Reply:
As stated in the PR description, let's focus on only supporting patterns that the OpenAPI compiler will generate for OpenAPI spec. For gRPC transcoding errors, we can handle them on a case-by-case basis later.

OpenAPI compiler will primarily generate:
- /foo/{x}
- /foo/**

This PR handles both of them. It also handles some other common cases like:
- /foo/*
- /foo/{x=**}

It will definitely not work for:
- {x=/foo/bar}
- {x=/foo/*/bar/**}

I think that's fine.